text/0000-project-toml-converter.md

```diff
@@ -0,0 +1,171 @@
+
+Example Project.TOML with 0.1 schema
+```
+[project]
+id = "<string>" # machine readable
+name = "<string>" # human readable
+version = "<string>"
+authors = ["<string>"]
+documentation-url = "<url>"
+source-url = "<url>"
+
+[[project.licenses]]
+type = "<string>"
+uri = "<uri>"
+
+[build]
+include = ["<string>"]
+exclude = ["<string>"]
+[[build.buildpacks]]
+id = "<string>"
+version = "<string>"
+uri = "<string>"
+[[build.env]]
+name = "<string>"
+value = "<string>"
+[metadata]
+# additional arbitrary keys allowed
+```
+
+
+Example Project.TOML with 0.2 schema
+```
+[_]
+id = "io.buildpacks.my-app"
+version = "0.1"
+
+[_.metadata]
+cdn = "https://cdn.example.com"
+
+[[_.metadata.assets]]
+url = "https://cdn.example.com/assets/foo.jar"
+checksum = "3b1b39893d8e34a6d0bd44095afcd5c4"
+
+buzz = ["a", "b", "c"]
+
+[io.buildpacks]
+builder = "cnbs/sample-builder:bionic"
+include = [
+    "cmd/",
+    "go.mod",
+    "go.sum",
+    "*.go"
+]
+
+[[io.buildpacks.group]]
+id = "io.buildpacks/java"
+version = "1.0"
+
+[[io.buildpacks.group]]
+id = "io.buildpacks/nodejs"
+version = "1.0"
+
+[[io.buildpacks.group]]
+id = "example/post-build"
+  
+    [io.buildpacks.group.script]
+    api = "0.5"
+    inline = "./post-build.sh"
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+- Another binary to maintain
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Having the lifecycle process project.toml directly (like prepare phase see: #555). This RFC is only focusing on the upgrade problem so that it can be done quickly.
+- Do nothing, hope everything will be fine (we can leave as it is). We think it's too hard for platform operators to support multiple versions of project.toml.
+- Having the translator binary translate from e.g., project.toml schema 0.1 -> project.toml schema 0.2. The platform API already represents the contract between the lifecycle and the platform, so it makes more sense than "project descriptor schema -> project descriptor schema" translation.. 
+
+# Prior Art
+[prior-art]: #prior-art 
+
+Discuss prior art, both the good and bad.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- What will the platform schema for project.toml look like?
+- Out of scope: parts of project.toml that aren't relevant to the platform api (like setting env vars)
+
+# Spec. Changes (OPTIONAL)
+The new platform schema could look something like:
+```
+[project]
+id = "<string>"
+version = "<string>"
+
+[build]
+builder = "cnbs/sample-builder:bionic"
+include = ["<string>"]
+exclude = ["<string>"]
+
+[[build.buildpacks]]
+id = "io.buildpacks/java"
+version = "1.0"
+
+[metadata]
+# all metadatas here
+```
+
+*Here are the namespace conversions*
+
+For conversion to/from v0.1
+```
+project = project
+build = build
+build.buildpacks = build.buildpacks
+metadata = metadata
+```
+
+For conversion to/from v0.2:
+
+```
+project = _
+build = io.buildpacks
+build.buildpacks = io.buildpacks.group
+metadata = _.metadata
```

Comment:
what happens to overlapped properties? ie. translating the following from `v0.2` to `v0.1`:

```toml
[_]
name = "My App"

# in this schema version this is consider arbitrary/user data
[project]
name = "Client 2 Project"
```

Reply:
maybe we can add a global prefix or postfix for keys which not included this list ? (just an idea)

Reply:
5c2075d should help clarify the mapping from 0.1 <-> 0.2 ...though I wasn't sure what to do about `[build.buildpacks.script]`.

Reply:
IIRC, we were going to ignore (not map) previously non-existent properties. It doesn't make sense for platforms to expect properties that aren't associated with their specification.